File: Python/quicksort.py
t = int(input())
def quicksort(arr,low,high):
    
    if low>=high:
        return 
    pivot = partition(arr,low,high)
    quicksort(arr,low,pivot-1)
    quicksort(arr,pivot+1,high)

def partition(arr,low,high):
    
    mid = (low+high)//2
    
    temp = arr[high]
    arr[high] = arr[mid]
    arr[mid] = temp
    
    i = low
    for j in range(low,high):
        if arr[j]<=arr[high]:
            temp = arr[j]
            arr[j] = arr[i]
            arr[i] = temp
            i+=1
    #print(i)
    temp = arr[high]
    arr[high] = arr[i]
    arr[i] = temp
    
    return i
    
for i in range(t):       # no of test cases
    n = int(input())     # no of elements in array 
    arr = [int(x) for x in input().split()] #enter array with spaces
    quicksort(arr,0,n-1)       # like 27 3 5 1 8
    print(arr)

File: DSA/Data Structures/Sorting Algorithms/Quicksort.cpp
#include <bits/stdc++.h>
#include "random.h"
using namespace std;

int partition(vector<int> &x, int low, int high)
{
    int pivot = x[low], i = low + 1, j = high;

    do
    {
        while (x[i] <= pivot)
            i++;

        while (x[j] > pivot)
            j--;

        if (i < j)
            swap(x[i], x[j]);
    } while (i < j);

    swap(x[low], x[j]);
    return j;
}
void quickSort(vector<int> &x, int low, int high)
{
    if (low < high)
    {
        int pI = partition(x, low, high);
        quickSort(x, low, pI - 1);
        quickSort(x, pI + 1, high);
    }
}

int main()
{
    vector<int> v;
    populate(v,10);
    print(v,"Unsorted");
    quickSort(v,0,v.size()-1);
    print(v,"Sorted  ");

    return 0;
}
